Add blog post about CI speed ups (#2167)
home-assistant/developers.home-assistant: blog/2024-05-09-improving-ci-test-execution.md

---
author: Robert Resch
authorURL: https://github.com/edenhaus
authorImageURL: https://avatars.githubusercontent.com/u/26537646?s=96&v=4
title: "How we managed to speed up our CI to save 168+ days of execution time per month"
---

We got a taste for speed after [UV gave us back 215 compute hours a month](2024-04-03-build-images-with-uv.md). 
Our CI workflow gets triggered on each commit, and due to the high volume of contributions, it was triggered 6647 times in March 2024.
The full runs, where the whole test suite is being executed, take a long time.

> It turned out that the plugin, we used to split the tests into 10 groups, was inefficient, as each pytest job needed to discover all tests, even when the job only executed a subset of them.

Reviewer: ```suggestion
It turned out that the plugin, we used to split the tests into 10 groups, was inefficient. Each pytest job needed to discover all tests, even when the job intended to only executed a subset of them.

```

---
author: Robert Resch
authorURL: https://github.com/edenhaus
authorImageURL: https://avatars.githubusercontent.com/u/26537646?s=96&v=4
title: "How we managed to speed up our CI to save 168+ days of execution time per month"
---

We got a taste for speed after [UV gave us back 215 compute hours a month](2024-04-03-build-images-with-uv.md). 
Our CI workflow gets triggered on each commit, and due to the high volume of contributions, it was triggered 6647 times in March 2024.
The full runs, where the whole test suite is being executed, take a long time.

It turned out that the plugin, we used to split the tests into 10 groups, was inefficient. Each pytest job needed to discover all tests, even when the job intended to only executed a subset of them.

> Now we have a seperate job to discover all tests and split them into 10 groups. The 10 pytest jobs only need to discover and execute a subset of all tests, which leads to reduced execution of 3 hours on each full run.

Reviewer: ```suggestion
Now we have a separate job to discover all tests and split them into 10 groups. The 10 pytest jobs only need to execute a subset of all tests. Not doing full-discovery in each test runner saves us 3 hours on each full run!

```

---
author: Robert Resch
authorURL: https://github.com/edenhaus
authorImageURL: https://avatars.githubusercontent.com/u/26537646?s=96&v=4
title: "How we managed to speed up our CI to save 168+ days of execution time per month"
---

We got a taste for speed after [UV gave us back 215 compute hours a month](2024-04-03-build-images-with-uv.md). 
Our CI workflow gets triggered on each commit, and due to the high volume of contributions, it was triggered 6647 times in March 2024.
The full runs, where the whole test suite is being executed, take a long time.

It turned out that the plugin, we used to split the tests into 10 groups, was inefficient. Each pytest job needed to discover all tests, even when the job intended to only execute a subset of them.

Now we have a separate job to discover all tests and split them into 10 groups. The 10 pytest jobs only need to execute a subset of all tests. Not doing full-discovery in each test runner saves us 3 hours on each full run!

> 

Reviewer: ```suggestion
```mermaid
flowchart LR
    A[Full run] --> R{Distribute work}

    R -->|One| D[Discover what tests to run]
    R -->|Two| E[Discover what tests to run]
    R -->|Three| F[Discover what tests to run]

    D --> D2[Run tests]
    E --> E2[Run tests]
    F --> F2[Run tests]

    AN[Full run] --> RW[Discover what tests to run]

    RW --> RN{Distribute work}

    RN -->|One| D2N[Run tests]
    RN -->|Two| E2N[Run tests]
    RN -->|Three| F2N[Run tests]
```

```

---
author: Robert Resch
authorURL: https://github.com/edenhaus
authorImageURL: https://avatars.githubusercontent.com/u/26537646?s=96&v=4
title: "How we managed to speed up our CI to save 168+ days of execution time per month"
---

We got a taste for speed after [UV gave us back 215 compute hours a month](2024-04-03-build-images-with-uv.md). 
Our CI workflow gets triggered on each commit, and due to the high volume of contributions, it was triggered 6647 times in March 2024.
The full runs, where the whole test suite is being executed, take a long time.

It turned out that the plugin, we used to split the tests into 10 groups, was inefficient. Each pytest job needed to discover all tests, even when the job intended to only execute a subset of them.

Now we have a separate job to discover all tests and split them into 10 groups. The 10 pytest jobs only need to execute a subset of all tests. Not doing full-discovery in each test runner saves us 3 hours on each full run!

```mermaid

> flowchart LR
>     A[Full run] --> R{Distribute work}
> 
>     R -->|One| D[Discover what tests to run]
>     R -->|Two| E[Discover what tests to run]
>     R -->|Three| F[Discover what tests to run]
> 
>     D --> D2[Run tests]
>     E --> E2[Run tests]
>     F --> F2[Run tests]
> 
> 
>     AN[Full run] --> RW[Discover what tests to run]
> 
>     RW --> RN{Distribute work}
> 
>     RN -->|One| D2N[Run tests]
>     RN -->|Two| E2N[Run tests]
>     RN -->|Three| F2N[Run tests]

Reviewer: ```suggestion
flowchart LR
    A[Full run] --> R{Distribute work}

    R -->|One| D[Discover tests to run]
    R -->|Two| E[Discover tests to run]
    R -->|Three| F[Discover tests to run]

    D --> D2[Run tests]
    E --> E2[Run tests]
    F --> F2[Run tests]

    AN[Full run] --> RW[Discover tests to run]

    RW --> RN{Distribute work}

    RN -->|One| D2N[Run tests]
    RN -->|Two| E2N[Run tests]
    RN -->|Three| F2N[Run tests]

```
